python/src/scipp/compat/dict.py

```diff
@@ -0,0 +1,227 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+# Copyright (c) 2020 Scipp contributors (https://github.com/scipp)
+# @author Neil Vaytet
+
+from .. import detail
+from .. import _utils as su
+from .._scipp import core as sc
+
+import numpy as np
+from collections import defaultdict
+from itertools import product
+
+
+def to_dict(scipp_obj):
+    """
+    Convert a scipp object (Variable, DataArray or Dataset) to a python dict.
+
+    :param scipp_obj: A Variable, DataArray or Dataset to be converted to a
+                      python dict.
+    :type scipp_obj: Variable, DataArray, or Dataset
+    :return: A dict containing all the information necessary to fully define
+             the supplied scipp object.
+    :rtype: dict
+    """
+    if su.is_variable(scipp_obj):
+        return _variable_to_dict(scipp_obj)
+    elif su.is_data_array(scipp_obj):
+        return _data_array_to_dict(scipp_obj)
+    elif su.is_dataset(scipp_obj):
+        # TODO: This currently duplicates all coordinates that would otherwise
+        # be at the Dataset level onto the individual DataArrays. We are also
+        # manually duplicating all attributes, since these are not carried when
```

Comment:
Why not use code as `_data_array_to_dict`, but store data as a dict of variables? Seems less messy and more correct, avoiding the problems you discuss in the comment?